added support to spring amqp batch

kananindzya opened the pull request:
closes #1706
## What does this PR do?
<!--
Replace this comment with a description of what's being changed by this PR.
Please explain the WHAT: A clear and concise description of what (patterns used, algorithms implemented, design architecture, message processing, etc.)
Can be as simple as Fixes #123 if there's a corresponding issue.
-->

## Checklist
<!--
You only have to fill one category of checkboxes, depending on the type of the PR.
Please DO NOT remove any item, ~~striking through~~ those that do not apply.
Just ignore the checkboxes of categories that don't apply.
-->

- [X] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [X] I have added tests that prove my fix is effective or that my feature works

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1716/detail/PR-1716/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1716/detail/PR-1716/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1716/detail/PR-1716/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1716/detail/PR-1716/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #1716 opened`


* Reason: `The PR is not allowed to run in the CI yet`


* Start Time: 2021-03-26T09:43:34.334+0000


* Duration: 3 min 57 sec


* Commit: d8935b856154962ffcd189037ef7a662d4a43f12




#### Trends :test_tube:


![Image of Build Times](https://apm-ci.elastic.co/job/apm-agent-java/job/apm-agent-java-mbp/job/PR-1716/buildTimeGraph/png)



</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1716/detail/PR-1716/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Load a resource file from a shared library`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-java/pipelines/apm-agent-java-mbp/pipelines/PR-1716/runs/1/steps/87/log/?start=0">here</a></li>
<li>Description: <code>approval-list/elastic/apm-agent-java.yml</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-java/pipelines/apm-agent-java-mbp/pipelines/PR-1716/runs/1/steps/90/log/?start=0">here</a></li>
<li>Description: <code>githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-03-26T09:46:02.607Z]  > git --version # 'git version 2.17.1'
[2021-03-26T09:46:02.607Z] using GIT_ASKPASS to set credentials GitHub user @apmmachine User + Personal Access Token
[2021-03-26T09:46:02.609Z]  > git fetch --no-tags --progress -- https://github.com/elastic/apm-agent-java.git +refs/heads/*:refs/remotes/origin/* # timeout=100
[2021-03-26T09:46:03.439Z] Pruning obsolete local branches
[2021-03-26T09:46:03.971Z] Merging remotes/origin/master commit 17453513cfae35b02b278f44df01d63cd8e207bb into PR head commit d8935b856154962ffcd189037ef7a662d4a43f12
[2021-03-26T09:46:03.395Z]  > git config remote.origin.url https://github.com/elastic/apm-agent-java.git # timeout=10
[2021-03-26T09:46:03.401Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-03-26T09:46:03.415Z]  > git config remote.origin.url https://github.com/elastic/apm-agent-java.git # timeout=10
[2021-03-26T09:46:03.448Z] Fetching upstream changes from https://github.com/elastic/apm-agent-java.git
[2021-03-26T09:46:03.448Z] using GIT_ASKPASS to set credentials GitHub user @apmmachine User + Personal Access Token
[2021-03-26T09:46:03.450Z]  > git fetch --no-tags --progress --prune -- https://github.com/elastic/apm-agent-java.git +refs/pull/1716/head:refs/remotes/origin/PR-1716 +refs/heads/master:refs/remotes/origin/master # timeout=100
[2021-03-26T09:46:03.980Z]  > git config core.sparsecheckout # timeout=10
[2021-03-26T09:46:03.995Z]  > git checkout -f d8935b856154962ffcd189037ef7a662d4a43f12 # timeout=100
[2021-03-26T09:46:04.250Z]  > git remote # timeout=10
[2021-03-26T09:46:04.288Z] Merge succeeded, producing d8935b856154962ffcd189037ef7a662d4a43f12
[2021-03-26T09:46:04.289Z] Checking out Revision d8935b856154962ffcd189037ef7a662d4a43f12 (PR-1716)
[2021-03-26T09:46:04.259Z]  > git config --get remote.origin.url # timeout=10
[2021-03-26T09:46:04.265Z] using GIT_ASKPASS to set credentials GitHub user @apmmachine User + Personal Access Token
[2021-03-26T09:46:04.266Z]  > git merge 17453513cfae35b02b278f44df01d63cd8e207bb # timeout=10
[2021-03-26T09:46:04.278Z]  > git rev-parse HEAD^{commit} # timeout=10
[2021-03-26T09:46:04.293Z]  > git config core.sparsecheckout # timeout=10
[2021-03-26T09:46:04.297Z]  > git checkout -f d8935b856154962ffcd189037ef7a662d4a43f12 # timeout=100
[2021-03-26T09:46:08.356Z] Commit message: "added support to spring amqp batch"
[2021-03-26T09:46:08.356Z] First time build. Skipping changelog.
[2021-03-26T09:46:08.390Z]  > git --version # timeout=10
[2021-03-26T09:46:08.406Z]  > git --version # 'git version 2.17.1'
[2021-03-26T09:46:08.413Z] fatal: bad object fbaca13d6c1a3541e236a4e1001d3af762ea4ccb
[2021-03-26T09:46:09.193Z] Masking supported pattern matches of $JOB_GCS_BUCKET or $NOTIFY_TO
[2021-03-26T09:46:09.286Z] Timeout set to expire in 1 hr 0 min
[2021-03-26T09:46:09.331Z] The timestamps step is unnecessary when timestamps are enabled for all Pipeline builds.
[2021-03-26T09:46:09.752Z] [INFO] Number of builds to be searched 10
[2021-03-26T09:46:10.069Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2021-03-26T09:46:10.101Z] Running in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1716/src/github.com/elastic/apm-agent-java
[2021-03-26T09:46:10.150Z] [INFO] gitCheckout: Checkout SCM PR-1716 with some customisation.
[2021-03-26T09:46:10.193Z] [INFO] Override default checkout
[2021-03-26T09:46:10.275Z] Sleeping for 10 sec
[2021-03-26T09:46:20.301Z] The recommended git tool is: git
[2021-03-26T09:46:20.443Z] using credential 2a9602aa-ab9f-4e52-baf3-b71ca88469c7-UserAndToken
[2021-03-26T09:46:20.510Z] Cloning the remote Git repository
[2021-03-26T09:46:20.526Z] Cloning repository https://github.com/elastic/apm-agent-java.git
[2021-03-26T09:46:20.555Z]  > git init /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1716/src/github.com/elastic/apm-agent-java # timeout=10
[2021-03-26T09:46:20.573Z] Using reference repository: /var/lib/jenkins/.git-references/apm-agent-java.git
[2021-03-26T09:46:20.574Z] Fetching upstream changes from https://github.com/elastic/apm-agent-java.git
[2021-03-26T09:46:20.574Z]  > git --version # timeout=10
[2021-03-26T09:46:20.582Z]  > git --version # 'git version 2.17.1'
[2021-03-26T09:46:20.583Z] using GIT_ASKPASS to set credentials GitHub user @apmmachine User + Personal Access Token
[2021-03-26T09:46:20.584Z]  > git fetch --tags --progress -- https://github.com/elastic/apm-agent-java.git +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-03-26T09:46:21.132Z]  > git config remote.origin.url https://github.com/elastic/apm-agent-java.git # timeout=10
[2021-03-26T09:46:21.136Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-03-26T09:46:21.143Z]  > git config remote.origin.url https://github.com/elastic/apm-agent-java.git # timeout=10
[2021-03-26T09:46:21.152Z] Fetching upstream changes from https://github.com/elastic/apm-agent-java.git
[2021-03-26T09:46:21.152Z] using GIT_ASKPASS to set credentials GitHub user @apmmachine User + Personal Access Token
[2021-03-26T09:46:21.153Z]  > git fetch --tags --progress -- https://github.com/elastic/apm-agent-java.git +refs/pull/1716/head:refs/remotes/origin/PR-1716 +refs/heads/master:refs/remotes/origin/master # timeout=10
[2021-03-26T09:46:21.595Z] Checking out Revision d8935b856154962ffcd189037ef7a662d4a43f12 (origin/PR-1716)
[2021-03-26T09:46:21.794Z] Commit message: "added support to spring amqp batch"
[2021-03-26T09:46:21.813Z]  > git --version # timeout=10
[2021-03-26T09:46:21.829Z]  > git --version # 'git version 2.17.1'
[2021-03-26T09:46:21.834Z] fatal: bad object fbaca13d6c1a3541e236a4e1001d3af762ea4ccb
[2021-03-26T09:46:21.589Z]  > git rev-parse origin/PR-1716^{commit} # timeout=10
[2021-03-26T09:46:21.598Z]  > git config core.sparsecheckout # timeout=10
[2021-03-26T09:46:21.608Z]  > git checkout -f d8935b856154962ffcd189037ef7a662d4a43f12 # timeout=10
[2021-03-26T09:46:22.644Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2021-03-26T09:46:23.336Z] + git fetch https://****:****@github.com/elastic/apm-agent-java.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2021-03-26T09:46:24.010Z] Running in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1716/src/github.com/elastic/apm-agent-java/.git
[2021-03-26T09:46:24.154Z] Archiving artifacts
[2021-03-26T09:46:25.004Z] + git rev-parse HEAD
[2021-03-26T09:46:25.373Z] + git rev-parse HEAD
[2021-03-26T09:46:25.702Z] + git rev-parse origin/pr/1716
[2021-03-26T09:46:25.773Z] [INFO] githubEnv: Found Git Build Cause: pr
[2021-03-26T09:46:26.078Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-03-26T09:46:27.078Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/apm-agent-java/pulls/1716/reviews return 0 elements
[2021-03-26T09:46:27.160Z] [INFO] githubPrCheckApproved: Title: added support to spring amqp batch - User: kananindzya - Author Association: CONTRIBUTOR
[2021-03-26T09:46:27.535Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2021-03-26T09:46:27.535Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2021-03-26T09:46:27.635Z] [INFO] Let's stop build #1. The PR is not allowed to run in the CI yet
[2021-03-26T09:46:27.669Z] Sleeping for 5 sec
[2021-03-26T09:46:28.783Z] Stage "Build" skipped due to earlier failure(s)
[2021-03-26T09:46:29.015Z] Stage "Tests" skipped due to earlier failure(s)
[2021-03-26T09:46:29.286Z] Stage "Unit Tests" skipped due to earlier failure(s)
[2021-03-26T09:46:29.305Z] Stage "Smoke Tests 01" skipped due to earlier failure(s)
[2021-03-26T09:46:29.323Z] Stage "Smoke Tests 02" skipped due to earlier failure(s)
[2021-03-26T09:46:29.341Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2021-03-26T09:46:29.358Z] Stage "Javadoc" skipped due to earlier failure(s)
[2021-03-26T09:46:29.573Z] Failed in branch Unit Tests
[2021-03-26T09:46:29.588Z] Failed in branch Smoke Tests 01
[2021-03-26T09:46:29.603Z] Failed in branch Smoke Tests 02
[2021-03-26T09:46:29.616Z] Failed in branch Benchmarks
[2021-03-26T09:46:29.630Z] Failed in branch Javadoc
[2021-03-26T09:46:29.939Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2021-03-26T09:46:30.051Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2021-03-26T09:46:30.102Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2021-03-26T09:46:30.556Z] Running on Jenkins in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1716
[2021-03-26T09:46:30.727Z] [INFO] getVaultSecret: Getting secrets
[2021-03-26T09:46:30.946Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-03-26T09:46:32.039Z] + chmod 755 generate-build-data.sh
[2021-03-26T09:46:32.039Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1716/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1716/runs/1 ABORTED 177427
[2021-03-26T09:46:32.290Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1716/runs/1/steps/?limit=10000 -o steps-info.json
[2021-03-26T09:46:32.841Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1716/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-03-26T09:46:32.841Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-03-26T09:46:33.752Z] Retry 2/3 exited 22, retrying in 2 seconds...

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

eyalkoren commented on apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchIteratorWrapper.java:
```diff
@@ -0,0 +1,69 @@
+                return message;
+            }
+            MessageProperties messageProperties = message.getMessageProperties();
+            if (messageProperties == null) {
+                return message;
+            }
```
Why? We should create a transaction anyway - a root transaction if there is no `tracecontext` property, or child transaction if there is. We already do that (create a transaction when there is no property). The same goes for `SpringAmqpMessageListenerInstrumentation` - we should allow `null` properties.
Please fix `SpringAmqpTransactionHelper#createTransaction()` instead to allow for `null` properties (and treat the same as if there are properties without context property).

kananindzya replied:
fixed

eyalkoren reviewed (commented):
Nice! 👏 
A few minor comments, overall looks really good!
I am pushing a small fix that will allow testing.

kananindzya commented on apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchIteratorWrapper.java:
```diff
@@ -0,0 +1,69 @@
+
+        Message message = delegate.next();
+        try {
+            if (message == null) {
+                return message;
+            }
```
deleted

kananindzya commented on apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchHelper.java:
```diff
@@ -0,0 +1,8 @@
+
+import java.util.List;
+
+public interface MessageBatchHelper<MSG> {
+
+    List<MSG> wrapMessageBatchList(List<MSG> messageBatchList);
+}
```
deleted

kananindzya commented on apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/SpringBaseInstrumentation.java:
```diff
@@ -45,4 +56,15 @@
     }
+
+    static class BaseAdvice {
+        protected static final MessageBatchHelper messageBatchHelper;
```
moved to SpringAmqpBatchMessageListenerInstrumentation$MessageListenerContainerWrappingAdvice

kananindzya commented on apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/SpringAmqpTransactionHelper.java:
```diff
@@ -0,0 +1,14 @@
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public interface SpringAmqpTransactionHelper {
```
yes, this interface not required. Deleted

eyalkoren commented:
Thanks for applying the review comments ❤️ 

Please see my last commits - I only removed another interface, changed helper names, added to CHANGELOG and added a test for distributed tracing.

Please let me know if you disagree with any of this.

eyalkoren commented:
run elasticsearch-ci/docs

eyalkoren commented:
@kananindzya we removed our `AssignTo` annotations and migrated to use the new Byte Buddy `AssignReturned` annotations - see other instrumentations and please it can build and test pass.
Sorry about that 🙁

kananindzya commented:
> @kananindzya we removed our `AssignTo` annotations and migrated to use the new Byte Buddy `AssignReturned` annotations - see other instrumentations and please it can build and test pass. Sorry about that slightly_frowning_face

fixed

eyalkoren commented:
run elasticsearch-ci/docs
